crates/consensus-logic/src/fork_choice_manager.rs

```diff
@@ -289,19 +301,48 @@ fn forkchoice_manager_task_inner<D: Database, E: ExecEngineCtl>(
     engine: &E,
     mut fcm_rx: mpsc::Receiver<ForkChoiceMessage>,
     csm_ctl: &CsmController,
+    mut cupdate_rx: Receiver<Arc<ClientUpdateNotif>>,
 ) -> anyhow::Result<()> {
-    loop {
+    let mut notif_closed = false;
+    let mut fcm_closed = false;
+
+    while !notif_closed || !fcm_closed {
         if shutdown.should_shutdown() {
             warn!("received shutdown signal");
             break;
         }
+        match cupdate_rx.try_recv() {
+            Ok(update_notif) => {
+                handle_new_state(
+                    &mut state,
+                    update_notif.new_state(),
+                    update_notif.tsn_output(),
+                )?;
+            }
+            Err(broadcast::error::TryRecvError::Empty) => {}
+            Err(broadcast::error::TryRecvError::Closed) => {
+                notif_closed = true;
+            }
+            Err(broadcast::error::TryRecvError::Lagged(_)) => {
+                // TODO: what to do?
+                notif_closed = true;
+            }
+        }
```

Comment:
Having a FCM message for state update is not an option here because that's what was causing the issue in the first place.

Using the watch channel seems very reasonable. At first it looked like only the latest state is not sufficient because the code was using `SyncAction` from each state update. But turns out it was using sync action only to finalize the block, which can be done from `state.sync().finalized_blkid()`.

Reply:
@bewakes For now let's shift towards not sending a `NewState` message back to the FCM anymore, but only sending a message when finalizing blocks with the newly finalized tip, the epoch idx, and then whatever else it might need.  We can call this `EpochVerificationState` or something.  We'll shift the split apart more later, but this is enough for now.

Reply:
Does the epoch idx mean the "checkpoint" idx?

I have made use of the `StatusRx` struct that contains all the information as watch channel. Whenever there's a new FCM message(not new state), it now checks the watch channel to get the event idx and corresponding status, and then updates accordingly. I don't think we need to add a new struct now.

Reply:
I've been shifting to using the term "epoch" since it's a more general term, with "checkpoint" referring to the data structure and operation itself.  Otherwise we have to use phrases like "checkpoint period" to describe the range of blocks and it's more verbose.

Ideally we should shift towards having the FCM not actually having to be aware of the full CSM state, as that makes those parts more coupled.  It should just operate with something small that gets extracted from the CSM state with just the information it strictly needs to process blocks.